connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -329,9 +335,15 @@ private void processTargetStateChanges(Set<String> connectorTargetStateChanges)
     public void halt() {
         synchronized (this) {
             // Clean up any connectors and tasks that are still running.
-            log.info("Stopping connectors and tasks that are still assigned to the worker");
-            worker.stopConnectors();
-            worker.stopAndAwaitTasks();
+            log.info("Stopping connectors and tasks that are still assigned to this worker.");
+            List<Callable<Void>> callables = new ArrayList<>();
```

Comment:
We need a "join" type barrier in the end to make sure all the callables have completed (this includes start or stop of both connectors and tasks). If instead of calling `invokeAll` (which will take a collection of callables as argument), this barrier can be achieved only with a call to `shutdown` followed by a call to `awaitTermination`, both on the executor service. Which is fine, unless we want to reuse this pool of threads. 

Depending on how frequently we expect the bulk starts and stops to happen we should pick between the two options: 
1) Frequent: use `invokeAll` and leave the pool around for the whole execution. 
2) Infrequent: create pool in every call and use `shutdown` + `awaitTermination`.